vhdx/src/vhdx_header.rs

```diff
@@ -0,0 +1,536 @@
+    pub signature: u32,
+    pub checksum: u32,
+    pub entry_count: u32,
+}
+
+const REGION_SIZE: u64 = 64 * 1024; // Each header size if 64KB
+const REGION_SIGN: u32 = 0x6967_6572; // "regi"
+const REGION1_START: u64 = 192 * 1024; // Region 1 start in Bytes
+const REGION2_START: u64 = 256 * 1024; // Region 2 start in Bytes
+
+impl RegionTableHeader {
+    /// Reads the Region Table Header structure from a reference VHDX file
+    pub fn new(f: &mut File, start: u64) -> Result<RegionTableHeader> {
+        // Read the whole header in to a buffer. We will need if for calculating checksum
+        let mut buffer = [0; REGION_SIZE as usize];
+        f.seek(SeekFrom::Start(start))
+            .map_err(Error::ReadingRegionTableHeader)?;
+        f.read_exact(&mut buffer).map_err(Error::ReadingHeader)?;
+
+        let signature = LittleEndian::read_u32(&buffer[0..3]);
+        if signature != REGION_SIGN {
+            return Err(Error::InvalidRegionSign);
+        }
+
+        let checksum = LittleEndian::read_u32(&buffer[4..7]);
+        let new_checksum = match calculate_checksum(&mut buffer, size_of::<u32>()) {
+            Ok(r) => r,
+            Err(e) => {
+                return Err(e);
+            }
+        };
+
+        if checksum != new_checksum {
+            return Err(Error::InvalidChecksum(String::from("Region")));
+        }
+
+        let entry_count = LittleEndian::read_u32(&buffer[8..11]);
+
+        if entry_count > 2047 {
+            return Err(Error::InvalidEntryCount);
+        }
+
+        Ok(RegionTableHeader {
+            signature,
+            checksum,
+            entry_count,
+        })
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct RegionTableEntry {
+    pub guid: u128,
+    pub file_offset: u64,
+    pub length: u32,
+    pub required: u32,
+}
+const REGION1_ENTRY_START: u64 = REGION1_START + REGION_SIZE;
+const REGION2_ENTRY_START: u64 = REGION2_START + REGION_SIZE;
+const MAX_ENTRIES: u64 = 2047;
+const ENTRY_SIZE: usize = 32; // Each Entry is 32 Bytes
+const VHDX_REGION_ENTRY_REQUIRED: u32 = 1;
+
+const BAT_GUID: &'static str = "2DC27766-F623-4200-9D64-115E9BFD4A08";
+const MDR_GUID: &'static str = "8B7CA206-4790-4B9A-B8FE-575F050F886E";
+
+impl RegionTableEntry {
+    /// Reads one Region Entry from a Region Table index that starts from 0
+    pub fn new(buffer: &[u8]) -> Result<RegionTableEntry> {
+        let guid = LittleEndian::read_u128(&buffer[0..15]);
+        let file_offset = LittleEndian::read_u64(&buffer[16..23]);
+        let length = LittleEndian::read_u32(&buffer[24..27]);
+        let required = LittleEndian::read_u32(&buffer[28..31]);
+
+        Ok(RegionTableEntry {
+            guid,
+            file_offset,
+            length,
+            required,
+        })
+    }
+
+    pub fn collect_entries(
+        f: &mut File,
+        start: u64,
+        entry_count: u32,
+    ) -> Result<(
+        Option<RegionTableEntry>,
+        Option<RegionTableEntry>,
```

Comment:
This should not return Options, but instead the actual structure. At the end, this function checks if the entry has been found, otherwise it returns an error, that's why you know you can return the type directly without wrapping it into an Option.